backend/backend-standard-and-best-practices.md

```diff
@@ -0,0 +1,278 @@
+
+```json
+{
+  "id": 1,
+  "first_name": "My first name",
+  "last_name": "My last name",
+  "email": "My email"
+}
+```
+
+```json
+{
+  "data": [
+    {
+      "id": 1,
+      "first_name": "My first name",
+      "last_name": "My last name",
+      "email": "My email"
+    },
+    {
+      "id": 2,
+      "first_name": "My first name",
+      "last_name": "My last name",
+      "email": "My email"
+    },
+    {
+      "id": 3,
+      "first_name": "My first name",
+      "last_name": "My last name",
+      "email": "My email"
+    }
+  ]
+}
+```
+### 2.2- Output parameters structure
+
+If we need to return an object that represents a certain entity, for example an user, we should avoid to return it using a key in the body containing all its data; we should directly return the data in the body instead.
+
+```json
+{
+  "id": 1,
+  "first_name": "My first name",
+  "last_name": "My last name",
+  "email": "My email"
+}
+```
+
+instead of:
+
+```json
+{
+  "user": {
+    "id": 1,
+    "first_name": "My first name",
+    "last_name": "My last name",
+    "email": "My email"
+  }
+}
+```
+
+## 3- Response bodies
+
+### 3.1- Pagination
+
+All paginated lists requested to our APIs should be returned using the following structure:
+
+```json
+{
+ "page": [
+    {
+      "id": 4,
+      "first_name": "My first name",
+      "last_name": "My last name",
+      "email": "My email"
+    },
+    {
+      "id": 5,
+      "first_name": "My first name",
+      "last_name": "My last name",
+      "email": "My email"
+    },
+    {
+      "id": 6,
+      "first_name": "My first name",
+      "last_name": "My last name",
+      "email": "My email"
+    }
+ ],
+ "count": 3,
+ "limit": 3,
+ "offset": 0,
+ "total_pages": 4,
+ "total_count": 12,
+ "previous_page": 1,
+ "current_page": 2,
+ "next_page": 3,
+ "previous_page_link": "domain.com/users?page=1&limit=3",
+ "current_page_link": "domain/com/users?page=2&limit=3",
+ "next_page_link": "domain.com/users?page=3&limit=3",
+ "first_page_link": "domain.com/users?page=1&limit=3",
+ "last_page_link": "domain.com/users?page=4&limit=3",
+ "n_page_link": "domain.com/users?page=X&limit=3"
+}
+```
+
+Some clarifications about the pagination response:
+- The starting page number should always be 1.
+- When returning the first page, the *previous_page* property should be `NULL`.
+- The parameters with the `_link` suffix are suggested when using link-oriented pagination; otherwise, they should not be included in the response body.
+
+### 3.2- Error forwarding
+
+When informing a client of a request error, the response body should be returned using the following structure:
+
+```json
+{
+  "errors": [
+    { "code": "111", "message": "age must be an int" }
+    { "code": "112", "message": "email is mandatory" }
+  ],
+  "origin": "api-name",
+  "stack_trace": "...",
+  "timestamp": "2019-09-10 03:14:07"
```

Comment:
This should be an iso 8601 date like `2019-12-02T17:40:53.855Z`

Reply:
I agree, maybe we can add as a clarification why we are using 8601 (I'm not sure if this is part of the goal of this document 🤔 )
I think that it's the standard used in Wolox but besides that maybe to list that it have some benefits like: 
- It can represent date, time, date-time, duration and interval
- It's not ambiguous
- Represents instants
- It's human readable

@mpicollo-wolox @KevinTemes what do you think?

Reply:
I do like the idea of clarifying why we use 8601, but perhaps for the next version of this document, since we need to release this relatively soon. I will change the timestamp example though.